File: app/src/main/java/org/illegaller/ratabb/hishoot2i/di/compenent/AppComponent.java
package org.illegaller.ratabb.hishoot2i.di.compenent;

import dagger.Component;
import javax.inject.Singleton;
import org.illegaller.ratabb.hishoot2i.HishootApplication;
import org.illegaller.ratabb.hishoot2i.di.module.ApplicationModule;
import org.illegaller.ratabb.hishoot2i.di.module.SystemServiceModule;
import org.illegaller.ratabb.hishoot2i.model.tray.TrayModule;

@Singleton
@Component(modules = { ApplicationModule.class, TrayModule.class, SystemServiceModule.class })
public interface AppComponent extends AppGraph {
  final class Initializer {
    private Initializer() {
      throw new UnsupportedOperationException();
    }

    public static AppComponent init(HishootApplication app) {
      return DaggerAppComponent.builder()
          .applicationModule(new ApplicationModule(app))
          .systemServiceModule(new SystemServiceModule())
          .trayModule(new TrayModule(app))
          .build();
    }
  }
}

File: app/src/main/java/org/illegaller/ratabb/hishoot2i/AppComponent.kt
package org.illegaller.ratabb.hishoot2i

import android.app.Application
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjectionModule
import dagger.android.AndroidInjector
import org.illegaller.ratabb.hishoot2i.data.DataModule
import org.illegaller.ratabb.hishoot2i.ui.UiModule
import javax.inject.Singleton

@Singleton
@Component(
    modules = [
        AndroidInjectionModule::class,
        AppModule::class,
        DataModule::class,
        UiModule::class
    ]
)
interface AppComponent : AndroidInjector<App> {
    @Component.Builder
    interface Builder {
        @BindsInstance
        fun application(application: Application): Builder

        fun build(): AppComponent
    }

    override fun inject(app: App)
}